src/nnpack/nnlib.jl

```diff
@@ -8,40 +8,40 @@ const AA1 = Union{AA{2}, AA{3}, AA{4}, AA{5}}
+softmax!(x::A) where A<:AbstractVecOrMat{Float64} = softmax!(Float32.(x))
 
-softmax!(x::AbstractVecOrMat{Float32}) =
+softmax!(x::A) where A<:AbstractVecOrMat{Float32} =
     nnp_softmax_output(x, x, threadpool = shared_threadpool)
 
-softmax!(y::AbstractVecOrMat{Float64}, x::AbstractVecOrMat{Float64}) = Float64.(softmax!(Float32.(y), Float32.(x)))
+softmax!(y::A, x::A) where A<:AbstractVecOrMat{Float64} = softmax!(Float32.(y), Float32.(x))
 
-softmax!(y::AbstractVecOrMat{Float32}, x::AbstractVecOrMat{Float32}) =
+softmax!(y::A, x::A) where A<:AbstractVecOrMat{Float32} =
     nnp_softmax_output(x, y, threadpool = shared_threadpool)
 
-softmax(x::AbstractVecOrMat{Float64}) = Float64.(softmax(Float32.(x)))
+softmax(x::A) where A<:AbstractVecOrMat{Float64} = softmax(Float32.(x))
 
-softmax(x::AbstractVecOrMat{Float32}) =
+softmax(x::A) where A<:AbstractVecOrMat{Float32} =
     nnp_softmax_output(x, similar(x), threadpool = shared_threadpool)
 
-maxpool(x::AbstractArray{Float64, 4}, k; pad = map(_->0,k), stride = k) =
-    Float64.(maxpool(Float32.(x), k, pad = pad, stride = stride))
+maxpool(x::A, k; pad = map(_->0,k), stride = k) where A<:AbstractArray{Float64, 4} =
+    maxpool(Float32.(x), k, pad = pad, stride = stride)
 
-function maxpool(x::AA{4}, k; pad = map(_->0,k), stride = k)
+function maxpool(x::A, k; pad = map(_->0,k), stride = k) where A<:AA{4}
     pad_, stride_ = expand(Val{length(k)}, pad), expand(Val{length(k)}, stride)
     ((size(x, 1) - k[1] + 2 * pad_[1]) % stride_[1] == 0 && (size(x, 2) - k[2] + 2 * pad_[2]) % stride_[2] == 0) || error("Choose the stride, pad and kernel size properly")
     maxpool!(similar(x, pdims(size(x), k, pad_, stride_)), x, k, pad = pad_, stride = stride_)
 end
 
-maxpool!(y::AbstractArray{Float64, 4}, x::AbstractArray{Float64, 4}, k; pad = map(_->0,k), stride = k) =
-    Float64.(maxpool!(Float32.(y), Float32.(x), k, pad = pad, stride = stride))
+maxpool!(y::A, x::A, k; pad = map(_->0,k), stride = k) where A<:AbstractArray{Float64, 4} =
+    maxpool!(Float32.(y), Float32.(x), k, pad = pad, stride = stride)
 
-maxpool!(y::AA{4}, x::AA{4}, k; pad = map(_->0,k), stride = k) =
+maxpool!(y::A, x::A, k; pad = map(_->0,k), stride = k) where A<:AA{4} =
     nnp_max_pooling_output(x, y, k, padding = expand(Val{length(k)}, pad), stride = expand(Val{length(k)}, stride), threadpool = shared_threadpool)
 
-conv(x::AbstractArray{Float64, 4}, w::AbstractArray{Float64, 4}; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) =
-    Float64.(conv(Float32.(x), Float32.(w), pad = pad, stride = stride, dilation = dilation, algo = algo))
+conv(x::A, w::A; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) where A<:AbstractArray{Float64, 4} =
+    conv(Float32.(x), Float32.(w), pad = pad, stride = stride, dilation = dilation, algo = algo)
 
-function conv(x::AA{4}, w::AA{4}; pad = 0, stride = 1, dilation = 1, algo = UInt32(0))
+function conv(x::A, w::A; pad = 0, stride = 1, dilation = 1, algo = UInt32(0)) where A<:AA{4}
```

Comment:
You have to be careful with these kinds of changes because they are not semantically the same; this `conv()` method now only takes `conv(::AA{Float32}, ::AA{Float32})` or `conv(::AA{Float64}, ::AA{Float64})` but not `conv(::AA{Float32}, ::AA{Float64})`.  Is that what you intended?

Reply:
Yes, since the support for different types in conv hasn't been merged yet IIRC, I was avoiding making such a change in this PR. I believe it would be confusing to have different behavior in different places. Also the AA is short for AbstractArray{Float32}, I guess I should change it.